src/material/timepicker/timepicker-input.ts

```diff
@@ -0,0 +1,421 @@
+    this._timepickerSubscription?.unsubscribe();
+    this._localeSubscription.unsubscribe();
+  }
+
+  /** Gets the ID of the input's label. */
+  _getLabelId(): string | null {
+    return this._formField?.getLabelId() || null;
+  }
+
+  /** Handles clicks on the input or the containing form field. */
+  private _handleClick = (): void => {
+    this.timepicker().open();
+  };
+
+  /** Handles the `input` event. */
+  protected _handleInput(value: string) {
+    const currentValue = this.value();
+    const date = this._dateAdapter.parseTime(value, this._dateFormats.parse.timeInput);
+    const hasChanged = !this._dateAdapter.sameTime(date, currentValue);
+
+    if (!date || hasChanged || !!(value && !currentValue)) {
+      // We need to fire the CVA change event for all nulls, otherwise the validators won't run.
+      this._assignUserSelection(date, true);
+    } else {
+      // Call the validator even if the value hasn't changed since
+      // some fields change depending on what the user has entered.
+      this._validatorOnChange?.();
+    }
+  }
+
+  /** Handles the `blur` event. */
+  protected _handleBlur() {
+    const value = this.value();
+
+    // Only reformat on blur so the value doesn't change while the user is interacting.
+    if (value && this._isValid(value)) {
+      this._formatValue(value);
+    }
+
+    this._onTouched?.();
+  }
+
+  /** Handles the `keydown` event. */
+  protected _handleKeydown(event: KeyboardEvent) {
+    // All keyboard events while open are handled through the timepicker.
+    if (this.timepicker().isOpen()) {
+      return;
+    }
+
+    if (event.keyCode === ESCAPE && !hasModifierKey(event) && this.value() !== null) {
+      event.preventDefault();
+      this.value.set(null);
+      this._formatValue(null);
+    } else if ((event.keyCode === DOWN_ARROW || event.keyCode === UP_ARROW) && !this.disabled()) {
+      event.preventDefault();
+      this.timepicker().open();
+    }
+  }
+
+  /** Sets up the code that watches for changes in the value and adjusts the input. */
+  private _respondToValueChanges(): void {
+    effect(() => {
+      const value = this._dateAdapter.deserialize(this.value());
+      const wasValid = this._lastValueValid;
+      this._lastValueValid = this._isValid(value);
+
+      // Reformat the value if it changes while the user isn't interacting.
+      if (!this._hasFocus()) {
+        this._formatValue(value);
+      }
+
+      if (value && this._lastValueValid) {
+        this._lastValidDate = value;
+      }
+
+      // Trigger the validator if the state changed.
+      if (wasValid !== this._lastValueValid) {
+        this._validatorOnChange?.();
+      }
+    });
+  }
+
+  /** Sets up the logic that registers the input with the timepicker. */
+  private _registerTimepicker(): void {
+    effect(() => {
+      const timepicker = this.timepicker();
+      timepicker.registerInput(this);
+      timepicker.closed.subscribe(() => this._onTouched?.());
```

Comment:
Im rusty on my rxjs, but should these be unsubscribed at some point

Reply:
Normally we would, but outputs get completed automatically on destroy: https://github.com/angular/angular/blob/main/packages/core/src/authoring/output/output_emitter_ref.ts#L40